sky/spot/recovery_strategy.py

```diff
@@ -321,15 +324,40 @@ def _launch(self,
                     logger.error(
                         'Failure happened before provisioning. Failover '
                         f'reasons: {reasons_str}')
-                    if raise_on_failure:
-                        raise exceptions.ProvisionPrechecksError(
-                            reasons=reasons)
-                    return None
+                    raise exceptions.ProvisionPrechecksError(reasons=reasons)
                 logger.info('Failed to launch the spot cluster with error: '
                             f'{common_utils.format_exception(e)})')
+            except exceptions.FileMountError as e:
+                # If the file mounts fails, the failure can be due to:
+                # 1. Errors before the job submission, e.g., invalid file mounts
+                #   -- user changed the bucket content on the cloud manually
+                #   during the launch/recovery. We should not retry the launch,
+                #   as the error is not recoverable.
+                # 2. Preemption during the file mounts. We should be able to
+                #   find the cluster not in UP status, and we can retry the
+                #   launch.
+                #
+                # Pull the actual cluster status from the cloud provider to
+                # determine whether the cluster is preempted.
+                time.sleep(5)
+                (cluster_status,
+                 _) = backend_utils.refresh_cluster_status_handle(
+                     self.cluster_name,
+                     force_refresh_statuses=set(status_lib.ClusterStatus))
+                if cluster_status == status_lib.ClusterStatus.UP:
+                    # Case 1: Errors before the job submission. We should fail
+                    # fast and not retry the launch.
+                    # Note: the underlying file mounts will retry for the
+                    # network glitches, so we do not need to retry here.
+                    logger.info('User file mounts failed. Not retrying.\n'
+                                '  Detailed exception: '
+                                f'{common_utils.format_exception(e)}')
+                    raise
+                # Case 2: The cluster is preempted during the launch, retry
+                # the launch.
+                logger.info('The cluster is preempted during the launch. '
+                            'Retrying.')
```

Comment:
Q:

- Shall we print the cluster status too?
- Is it possible to have a non-preempted cluster reach here with status INIT? E.g., file mounting failed after max retries so the status should be INIT?